redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/YogaLayout.kt

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+package app.cash.redwood.layout.view
+
+import android.content.Context
+import android.view.View
+import android.view.View.MeasureSpec
+import android.view.ViewGroup
+import app.cash.redwood.yoga.MeasureCallback
+import app.cash.redwood.yoga.MeasureMode
+import app.cash.redwood.yoga.Node
+import app.cash.redwood.yoga.Size
+import kotlin.math.roundToInt
+
+internal class YogaLayout(context: Context) : ViewGroup(context) {
+  private val nodes = mutableMapOf<View, Node>()
+  val rootNode = Node()
+
+  var applyModifier: (Node, Int) -> Unit = { _, _ -> }
+
+  init {
+    rootNode.measureCallback = ViewMeasureCallback(this)
+    applyLayoutParams(rootNode, layoutParams)
+  }
+
+  override fun addView(child: View, index: Int, params: LayoutParams) {
+    // Nodes with children cannot have measure functions.
+    rootNode.measureCallback = null
+
+    super.addView(child, index, params)
+
+    val childNode = child.asNode()
+    nodes[child] = childNode
+    rootNode.children += childNode
+  }
+
+  override fun removeView(view: View) {
+    removeViewFromYogaTree(view)
+    super.removeView(view)
+  }
+
+  override fun removeViewAt(index: Int) {
+    removeViewFromYogaTree(getChildAt(index))
+    super.removeViewAt(index)
+  }
+
+  override fun removeViewInLayout(view: View) {
+    removeViewFromYogaTree(view)
+    super.removeViewInLayout(view)
+  }
+
+  override fun removeViews(start: Int, count: Int) {
+    for (i in start until start + count) {
+      removeViewFromYogaTree(getChildAt(i))
+    }
+    super.removeViews(start, count)
+  }
+
+  override fun removeViewsInLayout(start: Int, count: Int) {
+    for (i in start until start + count) {
+      removeViewFromYogaTree(getChildAt(i))
+    }
+    super.removeViewsInLayout(start, count)
+  }
+
+  override fun removeAllViews() {
+    val childCount = childCount
+    for (i in 0 until childCount) {
+      removeViewFromYogaTree(getChildAt(i))
+    }
+    super.removeAllViews()
+  }
+
+  override fun removeAllViewsInLayout() {
+    val childCount = childCount
+    for (i in 0 until childCount) {
+      removeViewFromYogaTree(getChildAt(i))
+    }
+    super.removeAllViewsInLayout()
+  }
+
+  private fun removeViewFromYogaTree(view: View) {
+    val childNode = nodes[view] ?: return
+    val owner = childNode.owner ?: return
+
+    owner.children -= childNode
+    nodes -= view
```

Comment:
I'm not sure when only the second null check would fail, but I think we probably still want to remove from the map regardless, right?

```diff
     val childNode = nodes[view] ?: return
+    nodes -= view
     val owner = childNode.owner ?: return

     owner.children -= childNode
-    nodes -= view
```

Also, `remove()` returns the mapped value avoiding the need to perform the hashed lookup twice.
```suggestion
    val childNode = nodes.remove(view) ?: return

    val owner = childNode.owner ?: return
    owner.children -= childNode
```